index-handlers/src/main/java/no/sikt/nva/nvi/index/utils/InstitutionReportGenerator.java

```diff
@@ -0,0 +1,129 @@
+package no.sikt.nva.nvi.index.utils;
+
+import static no.sikt.nva.nvi.common.utils.ExceptionUtils.getStackTrace;
+import static no.sikt.nva.nvi.index.model.report.InstitutionReportHeaders.INSTITUTION_REPORT_HEADERS;
+import static nva.commons.core.attempt.Try.attempt;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Stream;
+import no.sikt.nva.nvi.index.aws.SearchClient;
+import no.sikt.nva.nvi.index.model.document.NviCandidateIndexDocument;
+import no.sikt.nva.nvi.index.model.document.NviContributor;
+import no.sikt.nva.nvi.index.model.document.NviOrganization;
+import no.sikt.nva.nvi.index.model.search.CandidateSearchParameters;
+import no.sikt.nva.nvi.index.xlsx.ExcelWorkbookGenerator;
+import org.opensearch.client.opensearch.core.search.Hit;
+import org.slf4j.Logger;
+
+public class InstitutionReportGenerator {
+
+    private static final Logger logger = org.slf4j.LoggerFactory.getLogger(InstitutionReportGenerator.class);
+    private static final String REPORT_REJECTED_VALUE = "N";
+    private static final String REPORT_PENDING_VALUE = "?";
+    private static final String REPORT_APPROVED_VALUE = "J";
+    private static final String REPORT_DISPUTED_VALUE = "T";
+    private final SearchClient<NviCandidateIndexDocument> searchClient;
+    private final String year;
+    private final URI topLevelOrganization;
+
+    public InstitutionReportGenerator(SearchClient<NviCandidateIndexDocument> searchClient,
+                                      String year,
+                                      URI topLevelOrganization) {
+        this.searchClient = searchClient;
+        this.year = year;
+        this.topLevelOrganization = topLevelOrganization;
+    }
+
+    public ExcelWorkbookGenerator generateReport() {
+        var nviCandidates = fetchNviCandidates();
+        var data = nviCandidates.stream().flatMap(this::generateDataRows).toList();
+        return new ExcelWorkbookGenerator(INSTITUTION_REPORT_HEADERS, data);
+    }
+
+    private Stream<List<String>> generateDataRows(NviCandidateIndexDocument candidate) {
+        return generateReportRowsForCandidate(candidate).stream();
+    }
+
+    private List<NviCandidateIndexDocument> fetchNviCandidates() {
+        return attempt(() -> searchClient.search(buildSearchRequest())).orElseThrow()
+                   .hits()
+                   .hits()
+                   .stream()
+                   .map(Hit::source)
+                   .toList();
+    }
+
+    private CandidateSearchParameters buildSearchRequest() {
+        return CandidateSearchParameters.builder()
+                   .withYear(year)
+                   .withTopLevelCristinOrg(topLevelOrganization)
+                   .build();
+    }
+
+    private List<List<String>> generateReportRowsForCandidate(NviCandidateIndexDocument candidate) {
+        return attempt(() -> generateReportRowsForContributorAffiliations(candidate)).map(Stream::toList)
```

Comment:
The reason I'm mapping the `Stream` to a `List` here (and then back to a Stream again in `generateDataRows`) is to force its evaluation, triggering any exceptions. This is because I want to log the specific candidate id if it fails to generate report rows for a candidate

Reply:
c.f. off-channel comms